pricegraph/src/orderbook/order.rs

```diff
@@ -127,7 +125,8 @@ pub struct Order {
     /// The maximum capacity for this order, this is equivalent to the order's
     /// remaining sell amount. Note that orders are also limited by their user's
     /// balance.
-    pub amount: f64,
+    /// None means unlimited.
+    pub amount: Option<u128>,
```

Comment:
We could also use u128::MAX to signal unbounded. That saves space in the struct because there is no extra byte needed for the Option.
I originally used Option but now that I write this I think u128::MAX is better. Could also make amount private and use a pub function to do the subtraction which would enforce this invariant.

Reply:
`u128::MAX` is a reserved contract value for an unbounded amount. Maybe we should create a new `Amount(u128)` wrapper type to represent this?

Reply:
It has special meaning in `Element`. It had no special meaning in `Order` up until this point. When we create an `Order` we check for this.

Reply:
> It had no special meaning in Order up until this point. 

Its because we converted `u128::MAX` to `f64::INFINITY` when we created it. We need to keep this behaviour since unlimited orders shouldn't decrease their `amount` when they get partially filled (only user balance).

Reply:
The same way I'm checking for `Some(amount)` now I would change this to check for `u128::MAX`. Will push a new commit that has this soon so we can look at it.

Reply:
My suggestion for the wrapper type was just to have a nice abstraction over this type and avoid creating invalid values. For example, `Some(u128::MAX)` is an invalid value (just like `2.0.powi(129)` would be with how it is currently written). But yes, its not strictly necessary, just thought it would be a nice time to make the code better :smile:

Reply:
Looks like that isn't actually faster in benchmarks and Option<u128> is more clear so let's keep the option. Or could create a new enum Amount Limited Unlimited.

Reply:
My only issue is that `Some(u128::MAX)` is an invalid value, so we have to make `amount` private to make sure this invariant is upheld. That's why I suggested a wrapper type in order to make it clear that it is a unsigned integer with "infinity" semantics. Although the same can be achieved with making `Order::amount` private and having an accessor method to it, and upholding the `!= u128::MAX` invariant in the update method.

Reply:
So, reading the code, I agree and think that the `Option` looks really nice. My only complaint is that it isn't clear from outside of `Order` what `amount == None` means. I suggest that the `if let Some(amount) = &mut order.amount { *amount = amount.saturating_sub(fill_amount); }` logic be moved a method on the `Order` type, and that `amount` be private.

Reply:
Agree that using option here is practical but the None case is an implementation detail that we don't need to leak to users. Otherwise I'd advertise for our own Option like type that is either Unbounded or Some(u128).

Reply:
I've made an Amount enum. Still about as practical as the Option but obvious what it means.

Reply:
While this is easier to understand, I still don't like how it's possible to create invalid values like `Amount::Remaining(u128::MAX)`